lib/controllers/issueish-detail-controller.js

```diff
@@ -44,14 +68,112 @@ export class BareIssueishDetailController extends React.Component {
       return <div>Issue/PR #{this.props.issueishNumber} not found</div>; // TODO: no PRs
     }
 
+    this.checkoutOp = this.nextCheckoutOp();
+
     return (
       <IssueishDetailView
         repository={repository}
         issueish={repository.issueish}
+        checkoutOp={this.checkoutOp}
         switchToIssueish={this.props.switchToIssueish}
       />
     );
   }
+
+  nextCheckoutOp() {
+    const {repository} = this.props;
+    const {issueish} = repository;
+
+    if (issueish.__typename !== 'PullRequest') {
+      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'Cannot check out an issue');
+    }
+
+    if (this.props.isMerging) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Merge in progress');
+    }
+
+    if (this.props.isRebasing) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Rebase in progress');
+    }
+
+    if (this.state.checkoutInProgress) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Checking out...');
+    }
+
+    // Determine if we already have this PR checked out.
+
+    const headPush = this.props.branches.getHeadBranch().getPush();
+    const headRemote = this.props.remotes.withName(headPush.getRemoteName());
+
+    // (detect checkout from pull/### refspec)
+    const fromPullRefspec =
+      headRemote.getOwner() === repository.owner.login &&
+      headRemote.getRepo() === repository.name &&
+      headPush.getShortRemoteRef() === `pulls/${issueish.number}/head`;
+
+    // (detect checkout from head repository)
+    const fromHeadRepo =
+      headRemote.getOwner() === issueish.headRepository.owner.login &&
+      headRemote.getRepo() === issueish.headRepository.name &&
+      headPush.getShortRemoteRef() === issueish.headRefName;
+
+    if (fromPullRefspec || fromHeadRepo) {
+      return this.checkoutOp.disable(checkoutStates.CURRENT, 'Current');
```

Comment:
Should it say what it is? Like `Current pull request`.

![image](https://user-images.githubusercontent.com/378023/42431833-b00b453c-8382-11e8-999d-c82ccd6ee939.png)


Or `Checked out`:

![image](https://user-images.githubusercontent.com/378023/42431822-9bf1a9ce-8382-11e8-8544-b28a755afd79.png)

Reply:
:+1: I went with "Checked out" to be a little less jarring on the transition from "Current pull request." Happy to be talked into any other copy though :smile:

I've also put the other status-dependent text in a button title rather than on the button text directly - the button will read "Checkout" unless you already have the PR's branch checked out, in which case it will say "Checked out." My logic is that the first time you see it, it would be confusing to have this big disabled button labelled "Rebase in progress," because you'd have no way to know what operation you couldn't perform because a rebase is in progress... on the other hand, I'm not sure how many people would notice the hover text as a way to figure out why it's disabled at a particular point in time.